### PR TITLE
TGUI Ports: Force Event Menu

### PR DESCRIPTION
**Pull request by Tsar-Salat**
Ports:

DM Refactors
- https://github.com/tgstation/tgstation/pull/68055
- https://github.com/tgstation/tgstation/pull/68472

UI Upgrades
- https://github.com/tgstation/tgstation/pull/71750
- https://github.com/tgstation/tgstation/pull/72690

## About The Pull Request

This converts the admin force-event menu into a proper TGUI menu, and sorts all the events by type.

Additionally, we can append a description to each event which lets admins know what events do what, rather than having to codedive to find out.

Furthermore, whilst keeping the changes from #10379, admins can now silence all events notifying the players by unchecking a box.

## Why It's Good For The Game

I received positive feedback from admins for this. I decided to port this after a conversation regarding how poorly worded some of the events are.

This should erase all ambiguity.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

https://github.com/BeeStation/BeeStation-Hornet/assets/62388554/97d6aca1-5fac-443a-aaa9-c88786978d79

</details>

## Changelog
:cl: rkz, Salex08, Profkanos, tralezab
add: TGUI-ifies the force-event menu
code: improves midround event canceling code
admin: admins can now silence events when forcing them
/:cl:

### CODE REVIEW

**Comment by Tsar-Salat** [comment]
Gonna grab more overhauls

**Comment by Tsar-Salat** [comment]
Alright, works fine so far. 

Just gotta add categories & desc to all these, then we are set.

image

**Comment by JixS4v** [comment]
Another day, another banger

**Review by itsmeow — CHANGES_REQUESTED** [review]
Having a bunch of horizontal stacking tabs looks really bad. This is a prime case for vertical tabs, with a scrollbar if necessary.

Also, we have historically called this "Trigger Event". I would prefer if it stayed that way (most importantly the JS window's title)

File: tgui/packages/tgui/interfaces/ForceEvent.tsx [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+export const PanelOptions = (props, context) => {
+  const [searchQuery, setSearchQuery] = useLocalState(context, 'searchQuery', '');
+
+  const [announce, setAnnounce] = useLocalState(context, 'announce', true);
+
+  return (
+    <Stack width="240px">
+      <Stack.Item>
+        <Icon name="search" />
```
Comment (itsmeow) [inline tgui/packages/tgui/interfaces/ForceEvent.tsx:88]: this really needs `vertical-align: middle` which will align it to the input box vertically. Alternatively you can use `line-height`.

File: tgui/packages/tgui/interfaces/ForceEvent.tsx [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+      // remove events not being searched for, if a search is active
+      if (searchQuery && !event.name.toLowerCase().includes(searchQuery)) {
+        return false;
+      }
+      return true;
+    }),
+    EVENT_PAGE_ITEMS
+  );
+
+  const sectionTitle = searchQuery ? 'Searching...' : category.name + ' Events';
```
Comment (itsmeow) [inline tgui/packages/tgui/interfaces/ForceEvent.tsx]:
```suggestion
  const sectionTitle = searchQuery ? 'Search Results' : category.name + ' Events';
```

File: code/modules/admin/force_event.dm [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+/datum/force_event/ui_static_data(mob/user)
+	var/static/list/category_to_icons
+	if(!category_to_icons)
+		category_to_icons = list(
+			EVENT_CATEGORY_AI = "robot",
+			EVENT_CATEGORY_ANOMALIES = "cloud-bolt",
+			EVENT_CATEGORY_BUREAUCRATIC = "print",
+			EVENT_CATEGORY_ENGINEERING = "wrench",
+			EVENT_CATEGORY_ENTITIES = "ghost",
+			EVENT_CATEGORY_FRIENDLY = "face-smile",
```
Comment (itsmeow) [inline code/modules/admin/force_event.dm:40]: from what I saw in the testing evidence, this is not an icon that exists, since it wasn't showing up.

**Review by itsmeow — CHANGES_REQUESTED** [review]

File: tgui/packages/tgui/interfaces/ForceEvent.tsx (hunk 1 above)
Comment (Tsar-Salat) [reply tgui/packages/tgui/interfaces/ForceEvent.tsx:88]: Don't think vertical align functions in our typescript for some reason. Shows up as undefined.

I will try the line-height.

Comment (itsmeow) [reply tgui/packages/tgui/interfaces/ForceEvent.tsx:88]: Just put an inline tag.

`style={{'vertical-align': 'middle'}}`

Comment (itsmeow) [reply tgui/packages/tgui/interfaces/ForceEvent.tsx:88]: can also just use a flex instead.
```
<Flex>
  <Flex.Item>
    <Icon name="search" mr={1} />
  </Flex.Item>
  <Flex.Item grow={1}>
    <Input
      placeholder="Search..."
      fluid
      value={searchText}
      onInput={(_, value) => setSearchText(value)}
      onEnter={(_, value) => orbitMostRelevant(value)}
    />
  </Flex.Item>
</Flex>
```

there is also `style={{ 'align-items': 'center' }}`

File: code/modules/admin/force_event.dm (hunk 3 above)
Comment (Tsar-Salat) [reply code/modules/admin/force_event.dm:40]: update fontawesome so they appear in 866157b

Comment (itsmeow) [reply code/modules/admin/force_event.dm:40]: Does this break old icons? I know fontawesome likes renaming stuff bwtween versions so a little concerned

**Comment by llol111** [comment]
Actualy good TGUI change 
I love shitting on TGUI being added to evrying just because you can,
but this is considerably better and nicer then the menu we have right now

**Comment by itsmeow** [comment]
Can you please look through existing UIs with lots of fontawesome for missing icons? I am quite concerned about updating it, because they rename stuff a *lot* between versions.

File: tgui/packages/tgui/interfaces/ForceEvent.tsx (hunk 1 above)
Comment (Tsar-Salat) [reply tgui/packages/tgui/interfaces/ForceEvent.tsx:88]: Gonna be honest, i dont know how to use flex

**Comment by Tsar-Salat** [comment]
I dont know how to use flex, so I dont really know if I can address maints changes here.

This ones going in the can I guess

**Pull request closed without merging**